### PR TITLE
Debezium Oracle Source

### PR DESCRIPTION
**Pull request by dlg99**
### Motivation

Add Debezium Source for Oracle.

### Modifications

Added Debezium Source for Oracle DB.
Added integration test.
Oracle JDBC is not included into nar (licensing issue) so the end user will have to package it.
Dropping jars into the pulsar/lib directory is not enough (connector classpath doesn't get it) so for the integration test I ended up updating the nar file with `jar uf`.

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

This change added tests and can be verified as follows:

  - Added integration test

```
$ $ mvn test -f tests/pom.xml -DintegrationTestSuiteFile=pulsar-io-ora-source.xml -DintegrationTests -Dtest=org.apache.pulsar.tests.integration.io.sources.debezium.PulsarDebeziumOracleSourceTest  -DtestRetryCount=0
...
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.apache.pulsar.tests.integration.io.sources.debezium.PulsarDebeziumOracleSourceTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 445.938 s - in org.apache.pulsar.tests.integration.io.sources.debezium.PulsarDebeziumOracleSourceTest
[INFO]
[INFO] Results:
[INFO]
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Apache Pulsar :: Tests 2.9.0-SNAPSHOT:
[INFO]
[INFO] Apache Pulsar :: Tests ............................. SUCCESS [  1.948 s]
[INFO] Apache Pulsar :: Tests :: Integration .............. SUCCESS [07:34 min]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  07:36 min
```

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): no
  - The public API: no
  - The schema: no
  - The default values of configurations: no
  - The wire protocol: no
  - The rest endpoints: no
  - The admin cli options: no
  - Anything that affects deployment: no

#### For contributor

For this PR, do we need to update docs?

Need to add description of the new connector; TBD (or will create an issue later).
Need to document requirement to download and package Oracle JDBC driver.

### CODE REVIEW

**Comment by Anonymitaet** [comment]
@dlg99 many thanks for providing doc info!
When adding docs, you can take the following guides as references:
- https://pulsar.apache.org/docs/en/next/io-connectors/#debezium-mysql
- https://hub.streamnative.io/connectors/sqs-source/v2.8.0.5/

**Review by sijie — COMMENTED** [review]
before reviewing this, is there any license issues we should clear out?

**Comment by merlimat** [comment]
This is depending on Oracle JDBC driver https://mvnrepository.com/artifact/com.oracle.ojdbc/ojdbc8/19.3.0.0 that is licensed with "Oracle Free Use Terms and Conditions (FUTC)". It's not clear if this license is compatible with ASLv2.

**Comment by dlg99** [comment]
@sijie @merlimat I removed dependency on Oracle JDBC and updated `tests/docker-images/latest-version-image/Dockerfile` to download ojdbc jars for the integration test. Users will have to deal with this. 
According to https://issues.apache.org/jira/browse/LEGAL-526 FUTC is a no-go for Apache projects.

**Comment by eolivelli** [comment]
What about creating one dedicated CI job only for Oracle CDC ?

File: pulsar-io/debezium/oracle/src/main/resources/debezium-oracle-source-config.yaml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+tenant: "public"
+namespace: "default"
+name: "debezium-oracle-source"
+topicName: "debezium-oracle-topic"
+archive: "connectors/pulsar-io-debezium-oracle-2.9.0-SNAPSHOT.nar"
```
Comment (eolivelli) [inline pulsar-io/debezium/oracle/src/main/resources/debezium-oracle-source-config.yaml:24]: we are hard coding the version here, I am not sure that this version will be changed while doing the release

In the PostGres connector we do not have this line, probably it is to be removed

File: pulsar-io/debezium/oracle/src/main/resources/debezium-oracle-source-config.yaml [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  ## config for Oracle XE, docker image: https://github.com/MaksymBilenko/docker-oracle-12c
+  ## docker run -d -p 1521:1521 quay.io/maksymbilenko/oracle-12c
+  database.hostname: "localhost"
+  database.port: "1521"
+  database.user: "sysdba"
+  database.password: "oracle"
+  database.dbname: "XE"
+  database.server.name: "XE"
+
+  database.history.pulsar.service.url: "pulsar://127.0.0.1:6650"
```
Comment (eolivelli) [inline pulsar-io/debezium/oracle/src/main/resources/debezium-oracle-source-config.yaml:38]: this is also to be removed or commented out

File: pulsar-io/debezium/oracle/src/main/resources/debezium-oracle-source-config.yaml [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+name: "debezium-oracle-source"
+topicName: "debezium-oracle-topic"
+archive: "connectors/pulsar-io-debezium-oracle-2.9.0-SNAPSHOT.nar"
+
+parallelism: 1
+
+configs:
+  ## config for Oracle XE, docker image: https://github.com/MaksymBilenko/docker-oracle-12c
+  ## docker run -d -p 1521:1521 quay.io/maksymbilenko/oracle-12c
+  database.hostname: "localhost"
```
Comment (eolivelli) [inline pulsar-io/debezium/oracle/src/main/resources/debezium-oracle-source-config.yaml:31]: should we comment out these lines?

**Review by eolivelli — CHANGES_REQUESTED** [review]
great work
I left some comments about hardcoding 2.9.0-SNAPSHOT, PTAL

File: pulsar-io/debezium/oracle/src/main/resources/debezium-oracle-source-config.yaml (hunk 1 above)
Comment (dlg99) [reply pulsar-io/debezium/oracle/src/main/resources/debezium-oracle-source-config.yaml:24]: We do have that in multiple connectors, versions are obsolete, not sure why we have that
```
$ grep -rI "archive:" . | grep -v /target/ 
./debezium/oracle/src/main/resources/debezium-oracle-source-config.yaml:archive: "connectors/pulsar-io-debezium-oracle-2.9.0-SNAPSHOT.nar"
./debezium/mongodb/src/main/resources/debezium-mongodb-source-config.yaml:archive: "connectors/pulsar-io-debezium-mongodb-2.4.0-SNAPSHOT.nar"
./debezium/postgres/src/main/resources/debezium-postgres-source-config.yaml:archive: "connectors/pulsar-io-debezium-postgres-2.4.0-SNAPSHOT.nar"
./debezium/mysql/src/main/resources/debezium-mysql-source-config.yaml:archive: "connectors/pulsar-io-debezium-mysql-2.4.0-SNAPSHOT.nar"
```

Similarly, the config.yaml files have example configuration. 
I don't mind removing that (as a separate change) if we agree that's the right thing to do.
Or the release procedure/script needs update to patch the version in these configs.

File: pulsar-io/debezium/oracle/src/main/resources/debezium-oracle-source-config.yaml (hunk 2 above)
Comment (dlg99) [reply pulsar-io/debezium/oracle/src/main/resources/debezium-oracle-source-config.yaml:38]: see https://github.com/apache/pulsar/pull/11520#discussion_r695920596

File: pulsar-io/debezium/oracle/src/main/resources/debezium-oracle-source-config.yaml (hunk 3 above)
Comment (dlg99) [reply pulsar-io/debezium/oracle/src/main/resources/debezium-oracle-source-config.yaml:31]: see https://github.com/apache/pulsar/pull/11520#discussion_r695920596

**Review by merlimat — APPROVED** [review]
👍

**Comment by eolivelli** [comment]
@sijie are you okay with merging this PR now ? it looks like licensing issues have been addressed.

**Comment by dlg99** [comment]
ping @sijie

**Comment by lhotari** [comment]
@eolivelli @dlg99 The new integration test seems to be flaky. Here's one sample failure: 
https://github.com/apache/pulsar/pull/12027/checks?check_run_id=3591596786#step:13:27471
```
Error:  testDebeziumOracleDbSource(org.apache.pulsar.tests.integration.io.sources.debezium.PulsarDebeziumOracleSourceTest)  Time elapsed: 1,800.028 s  <<< FAILURE!
org.testng.internal.thread.ThreadTimeoutException: Method org.apache.pulsar.tests.integration.io.sources.debezium.PulsarDebeziumOracleSourceTest.testDebeziumOracleDbSource() didn't finish within the time-out 1800000
```

**Comment by Anonymitaet** [comment]
Hi @dlg99 for the doc side, I think we already documented "the requirement to download and package Oracle JDBC driver" here(https://pulsar.apache.org/docs/en/next/io-debezium-source/#example-of-oracle)?

**Comment by dlg99** [comment]
@Anonymitaet correct, see https://github.com/apache/pulsar/pull/12213 for the docs

**Pull request merged**